File: right_side_view.cpp
// TC: O(N) where N is the total no of nodes in the tree;
// SC: O(N/2)


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        queue<TreeNode*>q;
        vector<int>result;
        if(root==NULL)return result;
        q.push(root);
        while(!q.empty()){
            int sz = q.size();
            for(int i = 0; i < sz;i++){
            root = q.front();q.pop();
            if(i == sz-1)result.push_back(root->val);
            if(root->left)q.push(root->left);
            if(root->right)q.push(root->right);
            }
        }
        return result;              
    }
};

File: rightsideView.py
# Time Complexity: O(n) : n is the number of nodes in the tree
# Space Complexity: O(h) : H is the height of the Tree

# Because we process the left node before the right node, the right node is always the last to be processed.
# As the DFS proceeds, for a given level, the right-most node of the level is processed at the last, 
# So we append/overwrite the value in the Result List.

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        self.result = []
        self.dfs(root, 0)
        return self.result
    
    def dfs(self, root: Optional[TreeNode], level: int) -> None:

        if root is None:
            return
        
        if len(self.result) == level:
            self.result.append(-101)
        
        if self.result[level]:
            self.result[level] = root.val
        
        self.dfs(root.left, level + 1)
        self.dfs(root.right, level + 1)


        